File: Javascript/Q_12.js
/**
Write a function that takes in a non-empty array of distinct integers and an integer representing a target sum.
The function should find all triplets in the array that sum up to the target sum
and return a two-dimensional array of all these triplets. The numbers in each triplet
should be ordered in ascending order, and the triplets themselves
should be ordered in ascending order with respect to the numbers they hold.

If no three numbers sum up to the target sum,
the function should return an empty array.



**/


function findTriplets(arr, targetSum) {
    // Sort the input array
    arr.sort((a, b) => a - b);

    const triplets = [];
    // Iterate over the array with two pointers: left and right
    for (let i = 0; i < arr.length; i++) {
        let left = i + 1;
        let right = arr.length - 1;
        while (left < right) {
            // Calculate the current sum
            const currentSum = arr[i] + arr[left] + arr[right];
            // If the current sum is equal to the target sum, add the triplet to the result list
            if (currentSum === targetSum) {
                triplets.push([arr[i], arr[left], arr[right]]);
                left += 1;
                right -= 1;
            }
            // If the current sum is less than the target sum, move the left pointer to the right to increase the sum
            else if (currentSum < targetSum) {
                left += 1;
            }
            // If the current sum is greater than the target sum, move the right pointer to the left to decrease the sum
            else {
                right -= 1;
            }
        }
    }
    return triplets;
}

let arr= [12, 3, 1, 2, -6, 5, -8, 6]

let targetSum =0

console.log(findTriplets(arr,targetSum))

File: Python/Arrays/Q_12/Q_12.py
"""

Write a function that takes in a non-empty array of distinct
integers and an integer representing a target sum.
The function should find all triplets in the array
that sum up to the target sum and return a two-dimensional
array of all these triplets. The numbers in each triplet should
be ordered in ascending order, and the triplets themselves should
 be ordered in ascending order with respect to the numbers they hold.

If no three numbers sum up to the target sum, the function should return
an empty array.

Sample Input
array = [12, 3, 1, 2, -6, 5, -8, 6]
targetSum = 0
Sample Output
[[-8, 2, 6], [-8, 3, 5], [-6, 1, 5]]

"""

def find_tripplets(array, targetSum):


   array.sort()
   tripplet=[]


   for i in range(len(array)):
       L = i +1
       R = len(array)-1
       while L<R:
         sum = array[i]+ array[L] + array[R]
         if sum == targetSum:
            tripplet.append([array[i], array[L] ,array[R]])
            L += 1
            R -= 1

         elif sum < targetSum:
            L +=1
         else :
            R-=1
   return tripplet

arr= [12, 3, 1, 2, -6, 5, -8, 6]

targetSum =0

print(find_tripplets(arr,targetSum))



"""O(n)3 Solution
In this case, the time complexity is O(n^3) because the number of iterations of the outermost loop
(which has a time complexity of O(n)) is multiplied by the number of iterations of the middle loop 
(which also has a time complexity of O(n)), which is in turn multiplied by the number of iterations of the innermost loop
 (also O(n)). This results in a total time complexity of O(n^3).



"""
# def find_triplets(arr, target_sum):
#     triplets = []
#     for i in range(len(arr)):
#         for j in range(i+1, len(arr)):
#             for k in range(j+1, len(arr)):
#                 if arr[i] + arr[j] + arr[k] == target_sum:
#                     triplets.append([arr[i], arr[j], arr[k]])
#     return triplets